File: legacy/secrt-server/cmd/secrt-server/reaper.go
package main

import (
	"context"
	"errors"
	"log/slog"
	"time"
)

const (
	expiryReaperInterval      = 5 * time.Minute
	expiryReaperDeleteTimeout = 10 * time.Second
)

type expiryReaperStore interface {
	DeleteExpired(ctx context.Context, now time.Time) (int64, error)
}

func runExpiryReaper(
	ctx context.Context,
	logger *slog.Logger,
	store expiryReaperStore,
	interval time.Duration,
	now func() time.Time,
) {
	if logger == nil {
		logger = slog.Default()
	}
	if now == nil {
		now = time.Now
	}
	if interval <= 0 {
		logger.Error("expiry reaper disabled: interval must be positive", "interval", interval)
		return
	}

	// Run once at startup so long-lived processes do not wait an entire tick
	// before purging expired rows.
	runExpiryReaperOnce(ctx, logger, store, now)

	ticker := time.NewTicker(interval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			runExpiryReaperOnce(ctx, logger, store, now)
		}
	}
}

func runExpiryReaperOnce(
	ctx context.Context,
	logger *slog.Logger,
	store expiryReaperStore,
	now func() time.Time,
) {
	if ctx.Err() != nil {
		return
	}

	cctx, cancel := context.WithTimeout(ctx, expiryReaperDeleteTimeout)
	defer cancel()

	deleted, err := store.DeleteExpired(cctx, now().UTC())
	if err != nil {
		// Shutdown/timeout cancellation is expected; avoid noisy logs.
		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
			return
		}
		logger.Error("expiry reaper delete failed", "err", err)
		return
	}
	if deleted > 0 {
		logger.Info("expired secrets deleted", "count", deleted)
	}
}

File: crates/secrt-server/src/reaper.rs
use std::sync::Arc;
use std::time::Duration;

use chrono::Utc;
use tokio::sync::oneshot;
use tracing::{error, info};

use crate::storage::SecretsStore;

pub const EXPIRY_REAPER_INTERVAL: Duration = Duration::from_secs(5 * 60);
pub const EXPIRY_REAPER_DELETE_TIMEOUT: Duration = Duration::from_secs(10);

pub fn start_expiry_reaper(store: Arc<dyn SecretsStore>) -> oneshot::Sender<()> {
    let (tx, mut rx) = oneshot::channel::<()>();

    tokio::spawn(async move {
        run_expiry_reaper_once(store.clone()).await;
        let mut ticker = tokio::time::interval(EXPIRY_REAPER_INTERVAL);

        loop {
            tokio::select! {
                _ = ticker.tick() => {
                    run_expiry_reaper_once(store.clone()).await;
                }
                _ = &mut rx => {
                    break;
                }
            }
        }
    });

    tx
}

pub async fn run_expiry_reaper_once(store: Arc<dyn SecretsStore>) {
    let timeout = tokio::time::timeout(
        EXPIRY_REAPER_DELETE_TIMEOUT,
        store.delete_expired(Utc::now()),
    )
    .await;

    match timeout {
        Ok(Ok(deleted)) => {
            if deleted > 0 {
                info!(count = deleted, "expired secrets deleted");
            }
        }
        Ok(Err(err)) => {
            error!(err = %err, "expiry reaper delete failed");
        }
        Err(_) => {
            error!("expiry reaper timed out");
        }
    }
}
